### PR TITLE
imp(PLM-477): Provide FOREACH for iterable data structure

### PR DESCRIPTION
**Pull request by kilikkuo**
[Why]
People might get confused because the validation pass when using LIST_OF but with a tuple, dict data.

[How]
1. Enforce LIST type check in LIST_OF
2. Use warning instead of print
3. Add more type hint
4. Update README
5. Add project url in setup.py
6. Revise deprecation message

[PLM-477]

Reason
--------------

Changes
--------------

Test Scope
--------------

Checks
--------------
- [x] Unit tests are included or not applicable.

[PLM-477]: https://hardcoretech.atlassian.net/browse/PLM-477?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: data_spec_validator/decorator/decorators.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
     from rest_framework.request import Request
 except ModuleNotFoundError as e:
     print(f'[DSV][WARNING] decorator: "dsv" cannot be used, {e}')
-from data_spec_validator.spec import validate_data_spec
+from data_spec_validator.spec import raise_if, validate_data_spec
```
Comment (xeonchen) [inline data_spec_validator/decorator/decorators.py:11]: Do you import this from `spec` and not from `spec.util` on purpose?

File: data_spec_validator/spec/actions.py [hunk 2]
```diff
@@ -70,5 +70,5 @@ def validate_data_spec(data, spec, **kwargs) -> bool:
 
     if not ok and not nothrow:
         error = _find_most_significant_error(failures)
-        raise error
+        raise_if(True, error)
```
Comment (xeonchen) [inline data_spec_validator/spec/actions.py]: Any benefit here?

**Review by xeonchen — APPROVED** [review]
It would be easier to review if refactor part and implementation part are separated. ;-)

File: data_spec_validator/decorator/decorators.py (hunk 1 above)
Comment (kilikkuo) [reply data_spec_validator/decorator/decorators.py:11]: Yes.

File: data_spec_validator/spec/actions.py (hunk 2 above)
Comment (kilikkuo) [reply data_spec_validator/spec/actions.py]: 恩, 我覺得我有點因噎廢食惹 XDD
來把 `raise_if(True, ...` 的拿掉, thanks

**Pull request merged**